pulp_container/app/registry_api.py

```diff
@@ -378,36 +383,136 @@ def get(self, request):
+        Prepare the paginated container _catalog response.
+        """
+        headers = {}
+        repositories_names = [repo["base_path"] for repo in data]
+        if repositories_names:
+            parsed_url = list(urlparse(self.url))
+            query_params = parse_qs(parsed_url[4])
+            query_params["n"] = str(self.n)
+            query_params["last"] = repositories_names[-1]
+            parsed_url[4] = urlencode(query_params)
+            url = urlunparse(parsed_url)
+            headers["Link"] = f'<{url}>; rel="next"'
+        return Response(headers=headers, data={"repositories": repositories_names})
+
+    page_size = api_settings.PAGE_SIZE
+
+
+class CatalogView(ContainerRegistryApiMixin, ListAPIView):
     """
     Handles requests to the /v2/_catalog endpoint
     """
 
-    def get(self, request):
-        """Handles GET requests for the /v2/_catalog endpoint."""
-        repositories_names = models.ContainerDistribution.objects.values_list(
-            "base_path", flat=True
-        )
-        return Response(data={"repositories": list(repositories_names)})
+    queryset = models.ContainerDistribution.objects.all().only("base_path")
+    serializer_class = ContainerCatalogSerializer
+    pagination_class = ContainerCatalogPagination
 
 
-class TagsListView(ContainerRegistryApiMixin, APIView):
+class ContainerTagListSerializer(ModelSerializer):
+    """
+    Serializer for Tags in the tags list endpoint of the registry.
+    """
+
+    class Meta:
+        model = models.Tag
+        fields = ["name"]
+
+
+class ContainerTagListPagination(BasePagination):
+    """
+    Pagination class to paginate tags by names according to the registry api specification.
+    """
+
+    def paginate_queryset(self, queryset, request, view=None):
+        """
+        Analyse the pagination parameters and prepare the queryset.
+        """
+        try:
+            self.n = int(request.query_params.get("n"))
+        except Exception:
+            self.n = self.page_size
+        else:
+            if self.n > 10 * self.page_size:
+                self.n = 10 * self.page_size
+        last = request.query_params.get("last")
+        self.url = request.build_absolute_uri()
+        self.path = request.resolver_match.kwargs["path"]
+
+        if last:
+            queryset = queryset.filter(name__gt=last)
+        return queryset.order_by("name")[: self.n]
+
+    def get_paginated_response(self, data):
+        """
+        Prepare the paginated container _catalog response.
+        """
+        headers = {}
+        tag_names = [tag["name"] for tag in data]
+        if tag_names:
+            parsed_url = list(urlparse(self.url))
+            query_params = parse_qs(parsed_url[4])
+            query_params["n"] = str(self.n)
+            query_params["last"] = tag_names[-1]
+            parsed_url[4] = urlencode(query_params)
+            url = urlunparse(parsed_url)
+            headers["Link"] = f'<{url}>; rel="next"'
```

Comment:
This will still send the link header even when whole result is displayed.

there are 2 options:

1. check in len of the tags_names. 
```
 def get_paginated_response(self, data):
        """
        Prepare the paginated container _catalog response.
        """
        headers = {}
        tag_names = [tag["name"] for tag in data]
        if tag_names:
            parsed_url = list(urlparse(self.url))
            query_params = parse_qs(parsed_url[4])
            query_params["n"] = str(self.n)
            query_params["last"] = tag_names[-1]
            parsed_url[4] = urlencode(query_params)
            url = urlunparse(parsed_url)
            if len(tag_names) == self.n:
                headers["Link"] = f'<{url}>; rel="next"'
        return Response(headers=headers, data={"name": self.path, "tags": tag_names})
```
if there have been requests 120 tags out of 120 we will still produce one unnecessary link header which will lead to to empty list. We can live with that alas.
2. If we want to eliminate even this conrner case we could store a result of the rest of the qs self.count_rest_of_qs = queryset.order_by("name")[self.n: ].count()` and compare the len of tag_names with value of this var, if 0 means we have returned full data. Might be performance not optimal :)

Reply:
I like option one. Even in option 2, there could be one more repo that gets deleted between the two calls and we'd return the empty list. So the client must be prepared for this event (getting an empty list) anyway.

Reply:
concur :) i decided to list 2nd option for the sake of completeness.